runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WindmillStateInternals.java

```diff
@@ -1999,56 +2005,79 @@ public Iterable<Entry<K, V>> read() {
           Future<Iterable<Entry<ByteString, Iterable<V>>>> persistedData = getFuture(false);
           try (Closeable scope = scopedReadState()) {
             Iterable<Entry<ByteString, Iterable<V>>> entries = persistedData.get();
-            Map<Object, ConcatIterables<V>> entryMap = Maps.newHashMap();
-            entries.forEach(
-                entry -> {
-                  try {
-                    final K key = keyCoder.decode(entry.getKey().newInput());
-                    final Object structuralKey = keyCoder.structuralValue(key);
-                    KeyState keyState =
-                        keyStateMap.computeIfAbsent(structuralKey, k -> new KeyState(key));
-                    if (keyState.existence == KeyExistence.KNOWN_NONEXISTENT) return;
-                    entryMap.compute(
-                        structuralKey,
-                        (k, v) -> {
-                          if (v == null) v = new ConcatIterables<>();
-                          v.extendWith(entry.getValue());
-                          keyState.existence = KeyExistence.KNOWN_EXIST;
-                          return v;
-                        });
-                  } catch (IOException e) {
-                    throw new RuntimeException(e);
-                  }
-                });
+            // If a key returned by windmill is known to be no longer exist or is completely cached
+            // locally, we can safely ignore the content of this key in windmill. Thus, we filter
+            // entries to filteredEntries which contains only keys that are known to exist and not
+            // fully cached.
+            Iterable<Entry<Object, Iterable<V>>> filteredEntries =
+                Iterables.filter(
+                    Iterables.transform(
+                        entries,
+                        entry -> {
+                          try {
+                            final K key = keyCoder.decode(entry.getKey().newInput());
+                            final Object structuralKey = keyCoder.structuralValue(key);
+                            KeyState keyState =
+                                keyStateMap.computeIfAbsent(structuralKey, k -> new KeyState(key));
+                            if (keyState.existence == KeyExistence.UNKNOWN_EXISTENCE) {
+                              keyState.existence = KeyExistence.KNOWN_EXIST;
+                            }
+                            return new AbstractMap.SimpleEntry<>(structuralKey, entry.getValue());
+                          } catch (IOException e) {
+                            throw new RuntimeException(e);
+                          }
+                        }),
+                    entry -> {
+                      KeyState keyState = keyStateMap.get(entry.getKey());
```

Comment:
can you combine the filter+transform to avoid looking up the key twice?
one idea would be to return null from transform if you want to filter, and then filter to remove nulls. But perhaps there is a better way to do with the stream apis, I'm not super familiar with them.